### PR TITLE
BUG: to_datetime re-parsing Arrow-backed objects

### LINKED ISSUE #52545: PERF: to_datetime doesn't recognise Arrow date time dtypes and converts them again
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

```
In [3]: dr = pd.date_range("2019-12-31", periods=1_000_000, freq="s").astype(pd.ArrowDtype(pa.timestamp(unit="ns")))

In [4]: %timeit pd.to_datetime(dr)
1.84 s ± 8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

this should be instant and not take close to 2 seconds

cc @MarcoGorelli 

### Installed Versions

<details>

main

</details>

### Prior Performance

_No response_

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #52545
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Ideally, we could push this down into DTA/DTI, but that doesn't seem to support Arrow arrays ATM.

### CODE REVIEW

File: pandas/tests/tools/test_to_datetime.py [hunk 1]
```diff
@@ -933,6 +933,20 @@ def test_to_datetime_dtarr(self, tz):
         result = to_datetime(arr)
         assert result is arr
 
+    @td.skip_if_no("pyarrow")
+    @pytest.mark.parametrize("utc", [True, False])
+    @pytest.mark.parametrize("tz", [None, "US/Central"])
+    def test_to_datetime_arrow(self, tz, utc):
+        import pyarrow as pa
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_datetime.py:940]:
```suggestion
        pa = pytest.importorskip("pyarrow")
```

Then you can remove `@td.skip_if_no("pyarrow")`

File: pandas/core/tools/datetimes.py [hunk 2]
```diff
@@ -400,6 +403,18 @@ def _convert_listlike_datetimes(
             arg = arg.tz_convert(None).tz_localize("utc")
         return arg
 
+    elif isinstance(arg_dtype, ArrowDtype) and np.issubdtype(
```
Comment (mroeschke) [inline pandas/core/tools/datetimes.py:406]: Could you use `arg_dtype.kind == "M"` instead of `np.issubdtype`?

**Review by mroeschke — COMMENTED** [review]
Could use a whatsnew too.

Also, does something similar happen to `to_timedelta`?

File: pandas/tests/tools/test_to_datetime.py (hunk 1 above)
Comment (lithomas1) [reply pandas/tests/tools/test_to_datetime.py]: Thanks, updated.

File: pandas/core/tools/datetimes.py (hunk 2 above)
Comment (lithomas1) [reply pandas/core/tools/datetimes.py]: Thx.

File: pandas/core/tools/datetimes.py [hunk 3]
```diff
@@ -400,6 +403,16 @@ def _convert_listlike_datetimes(
...
         return arg
 
+    elif isinstance(arg_dtype, ArrowDtype) and arg_dtype.kind == "M":
+        # TODO: Combine with above if DTI/DTA supports Arrow timestamps
+        if utc:
+            import pyarrow as pa
+
+            # array attribute has to exist since arg is Index/Series at this point
+            arg_arr = arg.array._pa_array
+            arg._pa_array = arg_arr.cast(pa.timestamp(arg_arr.type.unit, "utc"))
```
Comment (jbrockmendel) [inline pandas/core/tools/datetimes.py:413]: setting _pa_array looks really dangerous  (and weird; can't arg be a Series at this point)?  can we use a astype instead?

Comment (lithomas1) [reply pandas/core/tools/datetimes.py]: Thanks, updated.

**Comment by lithomas1** [comment]
> Also, does something similar happen to `to_timedelta`?

I think so, I'll try to get at it at a followup.

I think in the future, it might also make sense for DF inputs with Arrow backed ``year``, ``month``, etc. columns to convert to Arrow Backed dt columns.

e.g.
```
>>> import pandas as pd
+ /Users/thomasli/opt/miniconda3/envs/pandas-dev/bin/ninja
[1/1] Generating write_version_file with a custom command
>>> df = pd.DataFrame({'year': [2015, 2016],
...                    'month': [2, 3],
...                    'day': [4, 5]})
>>> df = df.astype("int64[pyarrow]")
>>> pd.to_datetime(df)
0   2015-02-04
1   2016-03-05
dtype: datetime64[ns]
```
should return something with dtype: ``timestamp[ns][pyarrow]``


Main thing I'm worried about with this PR is doing the wrong thing for ``utc=True`` (e.g. localizing instead of converting and vice versa).

(cc @phofl in case any thoughts and to double check my logic).

File: pandas/core/tools/datetimes.py [hunk 4]
```diff
@@ -400,6 +403,12 @@ def _convert_listlike_datetimes(
             arg = arg.tz_convert(None).tz_localize("utc")
         return arg
 
+    elif isinstance(arg_dtype, ArrowDtype) and arg_dtype.kind == "M":
+        # TODO: Combine with above if DTI/DTA supports Arrow timestamps
+        if utc:
+            arg = arg.astype("timestamp[ns, UTC][pyarrow]")
```
Comment (phofl) [inline pandas/core/tools/datetimes.py:409]: I'd rather use ``pd.ArrowDtype(pa...)`` here

**Review by phofl — COMMENTED** [review]
very small comment

File: pandas/core/tools/datetimes.py [hunk 5]
```diff
@@ -400,6 +403,12 @@ def _convert_listlike_datetimes(
             arg = arg.tz_convert(None).tz_localize("utc")
         return arg
 
+    elif isinstance(arg_dtype, ArrowDtype) and arg_dtype.kind == "M":
```
Comment (jbrockmendel) [inline pandas/core/tools/datetimes.py:406]: `pd.ArrowDtype(pa.date32())` also has kind == "M".  i think for this we need `pa.types.is_timestamp(arg_dtype.pyarrow_dtype)`.  or maybe `arg_dtype.type is Timestamp`

File: pandas/core/tools/datetimes.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/tools/datetimes.py:409]: .astype shouldn't allow converting a tz-naive to tzaware (or vice-versa).  i think this should use tz_convert/tz_localize instead

File: pandas/core/tools/datetimes.py [hunk 6]
```diff
@@ -403,6 +407,18 @@ def _convert_listlike_datetimes(
...
+    elif isinstance(arg_dtype, ArrowDtype) and arg_dtype.type is Timestamp:
+        # TODO: Combine with above if DTI/DTA supports Arrow timestamps
+        if utc:
+            # pyarrow uses UTC, not lowercase utc
+            if isinstance(arg, Index):
+                arg_array = cast(ArrowExtensionArray, arg.array)
+                arg = Index(arg_array._dt_tz_localize("UTC"))
+            else:
+                # ArrowExtensionArray
+                arg = arg._dt_tz_localize("UTC")
```
Comment (MarcoGorelli) [inline pandas/core/tools/datetimes.py:419]: is `tz_localize` always the right one, or do we need `tz_convert` if the input was timezone-aware to begin with?

In the test you've added, `dti_arrow` is always timezone-naive - would we make a timezone-aware example too please, just to check this?

File: pandas/core/tools/datetimes.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/tools/datetimes.py]: (btw if .astype works on ArrowEA to convert tznaive to tzaware we probably want to change that to match the non-arrow raising behavior)

Comment (lithomas1) [reply pandas/core/tools/datetimes.py]: > (btw if .astype works on ArrowEA to convert tznaive to tzaware we probably want to change that to match the non-arrow raising behavior)

Sure, I'll raise a PR after this one gets in.

File: pandas/core/tools/datetimes.py (hunk 6 above)
Comment (lithomas1) [reply pandas/core/tools/datetimes.py]: Thx for the catch.

I messed up the test (it does parametrize over UTC and US/Central), but I forgot to propagate the tz information in the tests.

File: pandas/tests/tools/test_to_datetime.py [hunk 7]
```diff
@@ -933,6 +933,32 @@ def test_to_datetime_dtarr(self, tz):
         result = to_datetime(arr)
         assert result is arr
 
+    # Doesn't work on Windows since tzpath not set correctly
+    @td.skip_if_windows
+    @pytest.mark.parametrize("arg_class", [Series, Index])
```
Comment (jbrockmendel) [inline pandas/tests/tools/test_to_datetime.py:938]: index_or_series fixture?

File: pandas/tests/tools/test_to_datetime.py [hunk 8]
```diff
@@ -933,6 +933,32 @@ def test_to_datetime_dtarr(self, tz):
...
+            pd.ArrowDtype(pa.timestamp(unit="ns", tz=tz if not utc else "UTC"))
+        )
+        if not utc and arg_class is not Series:
+            # Doesn't hold for utc=True, since that will astype
+            # to_datetime also returns a new object for series
+            assert result is dti_arrow
+        if arg_class is Series:
+            tm.assert_series_equal(result, expected)
+        else:
+            tm.assert_index_equal(result, expected, exact=False)
```
Comment (jbrockmendel) [inline pandas/tests/tools/test_to_datetime.py:960]: why do we need exact=False here?

File: pandas/tests/tools/test_to_datetime.py (hunk 7 above)
Comment (MarcoGorelli) [reply pandas/tests/tools/test_to_datetime.py:938]: tbh I'd prefer this, fixtures aren't great for readibility

File: pandas/tests/tools/test_to_datetime.py (hunk 8 above)
Comment (lithomas1) [reply pandas/tests/tools/test_to_datetime.py]: My bad, its' removed now.

File: pandas/tests/tools/test_to_datetime.py (hunk 7 above)
Comment (lithomas1) [reply pandas/tests/tools/test_to_datetime.py:938]: Leaving as is then.

**Review by MarcoGorelli — APPROVED** [review]
Well done, looks good to me!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lithomas1

**Comment by jbrockmendel** [comment]
Does the to_datetime docstring need to be updated to reflect non-DTI return?

**Pull request merged**